{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The idea is to get the genes(the genes that were translated, not intergenes) in the vicinity of the genes in the \"predicted_gene_blocks.fa\" file.\n",
    "\n",
    "![img](geneblocks.png)\n",
    "\n",
    "For example, in the above case, the red block is a gene block in our context and we have a modifier, an immunity and a trasnport gene there. We are taking all the genes that are in between them and also around them for a certain radius.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make the list of lists, where the members are the start and end number of \n",
    "gene blocks in the file predicted_gene_blocks.fa\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "os.chdir('/home/nafizh/try_Iddo_approaches/')\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os.path\n",
    "\n",
    "handle = open(\"predicted_gene_blocks.fa\", \"rU\")\n",
    "mydict = dict()\n",
    "myset = set()\n",
    "\n",
    "# taking out accession number, and their start and end,\n",
    "# and putting it into a dictionary where key is the \n",
    "# accession number and value is a list of lists with start,\n",
    "# and end numbers\n",
    "for record in SeqIO.parse(handle, \"fasta\") :\n",
    "    temp = record.id.split('|')\n",
    "    #print temp[0][10:20]\n",
    "    myset.add(temp[0][10:18])\n",
    "    if temp[0][10:18] in mydict:\n",
    "        mydict[temp[0][10:18]] = mydict[temp[0][10:18]] + [temp[2][6:], temp[3][4:]]\n",
    "    else:\n",
    "        mydict[temp[0][10:18]] = [temp[2][6:], temp[3][4:]]\n",
    "\n",
    "\n",
    "\n",
    "# so that the lists within the list are sorted\n",
    "# need them sorted because they are start and end of \n",
    "# protein sequence \n",
    "for key in mydict:\n",
    "    mydict[key].sort()\n",
    "\n",
    "print mydict['AE001437']\n",
    "print mydict['AE016877']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This code should get all the adjacent and in between genes of the genes \n",
    "predicted in the \"predicted_gene_blocks.fa\" file\n",
    "\"\"\"\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os.path\n",
    "\n",
    "os.chdir('/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction')\n",
    "\n",
    "def get_genes(input_file, output_file, my_key):\n",
    "    try:\n",
    "        \"\"\"get the desired gene blocks from .gbk files\"\"\"\n",
    "        mysequences = []\n",
    "        threshold = 1000\n",
    "        handle = open(input_file)\n",
    "\n",
    "        # the start position of the gene with lowest starting position\n",
    "        predicted_gene_block_start = mydict[my_key][0]\n",
    "        # the end position of the gene with highest ending position\n",
    "        predicted_gene_block_end = mydict[my_key][-1]\n",
    "\n",
    "\n",
    "        lower_thres_hold = int(predicted_gene_block_start) - threshold\n",
    "        upper_thres_hold = int(predicted_gene_block_end) + threshold\n",
    "\n",
    "        strand_dict = {-1: '-', 1 : '+'}\n",
    "\n",
    "        seq_record = SeqIO.parse(open(input_file), \"genbank\").next()\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                mystart = feature.location.start.position\n",
    "                myend = feature.location.end.position\n",
    "                #print mystart\n",
    "                #print myend\n",
    "                if lower_thres_hold <= mystart and myend <= upper_thres_hold and str(mystart + 1) not in mydict[my_key] and str(myend) not in mydict[my_key]:\n",
    "                    #print \"hello\"\n",
    "                    gene_seq = seq_record.seq[mystart : myend]\n",
    "                    mysequences.append(\n",
    "                          SeqRecord(gene_seq, id=\"%s:%d-%d%s\" % (seq_record.name,mystart + 1, myend, strand_dict[feature.strand]),\n",
    "                          description=\"%s %d-%d %s\" % (seq_record.name, mystart + 1, myend, strand_dict[feature.strand])))\n",
    "\n",
    "\n",
    "        SeqIO.write(mysequences, open(output_file, 'w'), \"fasta\")\n",
    "    except Exception as e:\n",
    "        print \"Error at\", input_file, e\n",
    "\n",
    "#get_genes('AE001437.gbk', 'rahi.fa', 'AE001437')\n",
    "\n",
    "root_dir = \"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction\"\n",
    "\n",
    "\"\"\"Going through all .gbk files and getting desired gene blocks from them\"\"\"\n",
    "for root, subFolders, files in os.walk(root_dir):\n",
    "    for fname in files:\n",
    "        genome_files = []\n",
    "        ext = os.path.splitext(os.path.basename(fname))[1]\n",
    "        name = os.path.splitext(os.path.basename(fname))[0]\n",
    "        if ext == \".gbk\":\n",
    "            output_file = name + \"_desired_genes.fa\"\n",
    "            absfile= os.path.join(root,fname)\n",
    "            my_key = fname.split('.')[0]\n",
    "            get_genes(absfile, output_file, my_key)\n",
    "            \n",
    "print \"Calculation done\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**After we get the \\*_desired_genes.fa files, we cat them all to a file called all_desired_genes.fa and then we  blast clust all the sequences on that file. We then translated all the sequences with transeq by the clusters we got from doing blast clust. We cat all those translations to a file called all_translations.fa and then blasted that file against the nr database. Finally, we got the blast result in the \"blast_res_for_all_translations.xml\" file. We then parsed it by the following code.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Getting only proteins that have some unknown proteins annotated to them\n",
    "# for blast result of all 777 protein sequences\n",
    "\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "result_handle = open(\"blast_res_for_all_translations.xml\")\n",
    "res_file = open(\"new_result_all_translations.fasta\", 'w')\n",
    "sec_file = open(\"new_result_all_translations1.fasta\", 'w')\n",
    "\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "\n",
    "track_l_all = []\n",
    "\n",
    "for i, blast_record in enumerate(blast_records):\n",
    "    for j, alignment in enumerate(blast_record.alignments): # j is basically the hit number within a blast record/blast_iteration\n",
    "        if j == 1: # only the top hit\n",
    "            break\n",
    "        for hsp in alignment.hsps:\n",
    "            if \"unknown function\" in alignment.title:\n",
    "                if i + 1 in track_l_all: # so that does not enter the same record/blast_iteration again\n",
    "                    continue\n",
    "                else:\n",
    "                    res_file.write(\">seq num: %d | hit num: %d | query: %s | hit-id: %s | hit-length: %d | hit-score: %d | e-value: %s|\\n\\n\" % (i + 1, j + 1, blast_record.query, alignment.title, alignment.length, hsp.score, (hsp.expect)))\n",
    "                    res_file.write(hsp.query[0:90] + \"...\\n\")\n",
    "                    res_file.write(hsp.match[0:90] + \"...\\n\")\n",
    "                    res_file.write(hsp.sbjct[0:90] + \"...\\n\")\n",
    "                    res_file.write(\"\\n\\n\")\n",
    "                    sec_file.write(\">seq num: %d | hit num: %d | query: %s | hit-id: %s | hit-length: %d | hit-score: %d | e-value: %s|\\n\\n\\n\" % (i+1, j + 1, blast_record.query, alignment.hit_id, alignment.length, hsp.score, (hsp.expect)))\n",
    "                    track_l_all = track_l_all + [i + 1]\n",
    "        \n",
    "\n",
    "            \n",
    "res_file.close()\n",
    "sec_file.close()\n",
    "print \"Calculation done\"   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**We have got all the bacteriocin hits with \"unknown function\". Now all the following code is to annotate them nicely with protein and dna sequences with their accession ids and species name.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Creating a temporary file of bacteriocins with unknown protein hits in a\n",
    "# format similar to all_translations.fa\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "\n",
    "bac_file = \"all_bacteriocins_with_unknown_protein_hit.fa\"\n",
    "cd_hit_handle = open(\"all_translations.fa\")\n",
    "\n",
    "mysequences = []\n",
    "\n",
    "j = 0\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(cd_hit_handle, \"fasta\")):\n",
    "    if i + 1 in track_l_all: # track_l_all comes from the previous cell\n",
    "        mysequences.append(SeqRecord(record.seq, id= (\"seq num: %d | serial num: %d |\") % (i + 1, j + 1), \n",
    "                                                         description= record.description))\n",
    "        j += 1\n",
    "    SeqIO.write(mysequences, open(bac_file, 'w'), \"fasta\")\n",
    "\n",
    "print \"Done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Creating a file with the desired putative bacteriocins with accession id, species name and co-ordinate number\n",
    "'''\n",
    "\n",
    "# Getting the accession ids as value in a dictionary where the key are the cordinates\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "\n",
    "bac_file = open(\"all_bacteriocins_with_unknown_protein_hit.fa\")\n",
    "all_desired_gene = open(\"all_desired_genes.fa\")\n",
    "\n",
    "coordinates = []\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(bac_file, \"fasta\")):\n",
    "    coordinates.append(record.description.split(\"|\")[2].split(\" \")[2])\n",
    "\n",
    "accessions = dict()\n",
    "for i, record in enumerate(SeqIO.parse(all_desired_gene, \"fasta\")):\n",
    "    #print record.description.split(\":\")[0]\n",
    "    for coord in coordinates:\n",
    "        if coord in record.description:\n",
    "            temp = record.description.split(\":\")[0]\n",
    "            accessions[coord] = temp\n",
    "\n",
    "print accessions\n",
    "            \n",
    "print len(accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Getting the genome ids for the accession ids\n",
    "\n",
    "from Bio import Entrez\n",
    "Entrez.email     = \"nafizh@iastate.edu\"\n",
    "\n",
    "accessions_list = []\n",
    "\n",
    "for value in accessions.itervalues():\n",
    "    accessions_list.append(value) \n",
    "\n",
    "print accessions_list\n",
    "print len(accessions_list)\n",
    "\n",
    "genomeAccessions1 = accessions_list[0:20]\n",
    "search1           = \" \".join(genomeAccessions1)\n",
    "print search1\n",
    "handle1           = Entrez.read(Entrez.esearch(db=\"nucleotide\", term=search1, retmode=\"xml\"))\n",
    "genomeIds1        = handle1['IdList']\n",
    "print genomeIds1\n",
    "print len(genomeIds1)\n",
    "\n",
    "# genomeAccessions2 = accessions_list[20:40]\n",
    "# search2           = \" \".join(genomeAccessions2)\n",
    "# print search2\n",
    "# handle2           = Entrez.read(Entrez.esearch(db=\"nucleotide\", term=search2, retmode=\"xml\"))\n",
    "# genomeIds2        = handle2['IdList']\n",
    "# print genomeIds2\n",
    "# print len(genomeIds2)\n",
    "# # # handle = Entrez.efetch(db=\"nucleotide\", id=\"186972394\", retmode=\"xml\")\n",
    "# # # record = Entrez.read(handle)\n",
    "# # # handle.close()\n",
    "# # # record[0][\"GBSeq_definition\"]\n",
    "\n",
    "# genomeAccessions3 = accessions_list[40:]\n",
    "# search3           = \" \".join(genomeAccessions3)\n",
    "# print search3\n",
    "# handle3           = Entrez.read(Entrez.esearch(db=\"nucleotide\", term=search3, retmode=\"xml\"))\n",
    "# genomeIds3        = handle3['IdList']\n",
    "# print genomeIds3\n",
    "# print len(genomeIds3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Creating a dicitionary where the key is the coordinate number and the values are the\n",
    "# respective genome ids and accession ids\n",
    "\n",
    "genome_ids = genomeIds1 \n",
    "#+ genomeIds2 + genomeIds3\n",
    "\n",
    "print genome_ids\n",
    "accession_dict = dict()\n",
    "\n",
    "print accessions\n",
    "\n",
    "track = 0\n",
    "for key, val in accessions.iteritems():\n",
    "    temp = [] + [val]\n",
    "    #accession_dict[key] = temp.append(genome_ids[track])\n",
    "    #print temp\n",
    "    temp1 = temp + [(str(genome_ids[track]))]\n",
    "    #print temp1\n",
    "    accession_dict[key] = temp1\n",
    "    track += 1\n",
    "    \n",
    "print accession_dict\n",
    "print len(accession_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Getting species name from the genome Ids\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "\n",
    "all_handle = open(\"all_translations.fa\")\n",
    "\n",
    "# dictionary where keys are coordinates and values are respective species names\n",
    "species_name = dict()\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(all_handle, \"fasta\")):\n",
    "    if i + 1 in track_l_all:\n",
    "        print record.description\n",
    "        my_key = record.description.split(\" \")[1]\n",
    "        my_id = accession_dict[my_key][1]\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id = my_id, retmode=\"xml\")\n",
    "        my_record = Entrez.read(handle)\n",
    "        species_name[my_key] = my_record[0][\"GBSeq_definition\"]\n",
    "        handle.close()\n",
    "        \n",
    "print species_name\n",
    "print len(species_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Writing the file of putative bacteriocins with protein sequences for Iddo\n",
    "output: \"10_bacteriocins_with_unknown_protein_hit_with_ids.fa\" file\n",
    "'''\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "\n",
    "bac_file = \"10_bacteriocins_with_unknown_protein_hit_with_ids.fa\"\n",
    "all_handle = open(\"all_translations.fa\")\n",
    "\n",
    "mysequences = []\n",
    "\n",
    "j = 0\n",
    "count = 0\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(all_handle, \"fasta\")):\n",
    "    if i + 1 in track_l_all:\n",
    "        print record.description\n",
    "        my_key = record.description.split(\" \")[1]\n",
    "        mysequences.append(SeqRecord(record.seq, id= (\"seq num: %d | serial num: %d | acession id: %s | species: %s|\") % (i + 1, j + 1, accession_dict[my_key][0],\n",
    "                             species_name[my_key]), description= record.description))\n",
    "        j += 1\n",
    "    SeqIO.write(mysequences, open(bac_file, 'w'), \"fasta\")\n",
    "\n",
    "print \"Done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Writing the file of putative bacteriocins with dna sequences for Iddo\n",
    "output : \"10_bacteriocins_with_unknown_protein_hit_with_ids_dna_seq.fa\" file\n",
    "'''\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "os.chdir(\"/home/nafizh/try_Iddo_approaches/all_gbk_files_in_prediction/desired_gene_blocks/translations\")\n",
    "\n",
    "bac_file = \"10_bacteriocins_with_unknown_protein_hit_with_ids_dna_seq.fa\"\n",
    "all_handle = open(\"all_translations.fa\")\n",
    "all_desired_gene = open(\"all_desired_genes.fa\")\n",
    "\n",
    "mysequences = []\n",
    "\n",
    "j = 0\n",
    "count = 0\n",
    "\n",
    "# dictionary where keys are coordinates and values are respective sequences\n",
    "seq_dict = dict()\n",
    "\n",
    "for gene_record in SeqIO.parse(all_desired_gene, \"fasta\"):\n",
    "    for key in accession_dict:\n",
    "        if key in gene_record.description:\n",
    "            seq_dict[key] = gene_record.seq\n",
    "\n",
    "print seq_dict\n",
    "print len(seq_dict)\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(all_handle, \"fasta\")):\n",
    "    if i + 1 in track_l_all:\n",
    "        print record.description\n",
    "        my_key = record.description.split(\" \")[1]\n",
    "        mysequences.append(SeqRecord(seq_dict[my_key], id= (\"seq num: %d | serial num: %d | acession id: %s | species: %s|\") % (i + 1, j + 1, accession_dict[my_key][0],\n",
    "                             species_name[my_key]), description= record.description))\n",
    "        j += 1\n",
    "    SeqIO.write(mysequences, open(bac_file, 'w'), \"fasta\")\n",
    "\n",
    "print \"Done\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
